Source/WebCore/page/NavigationHistoryEntry.cpp

```diff
@@ -40,14 +40,24 @@ namespace WebCore {
 
 WTF_MAKE_TZONE_OR_ISO_ALLOCATED_IMPL(NavigationHistoryEntry);
 
-NavigationHistoryEntry::NavigationHistoryEntry(ScriptExecutionContext* context, Ref<HistoryItem>&& historyItem)
+NavigationHistoryEntry::NavigationHistoryEntry(ScriptExecutionContext* context, Ref<HistoryItem>&& historyItem, String urlString, WTF::UUID key, WTF::UUID id)
     : ContextDestructionObserver(context)
-    , m_urlString(historyItem->urlString())
-    , m_id(WTF::UUID::createVersion4())
+    , m_urlString(urlString)
+    , m_key(key)
+    , m_id(id)
     , m_associatedHistoryItem(WTFMove(historyItem))
 {
 }
 
+Ref<NavigationHistoryEntry> NavigationHistoryEntry::create(ScriptExecutionContext* context, const NavigationHistoryEntry& other)
+{
+    Ref historyItem = other.m_associatedHistoryItem;
+    RefPtr state = historyItem->navigationAPIStateObject();
+    auto ret = adoptRef(*new NavigationHistoryEntry(context, WTFMove(historyItem), other.m_urlString, other.m_key, other.m_id));
+    ret->setState(WTFMove(state));
+    return ret;
```

Comment:
I did not pay much attention to this since I think we'll have to improve the state code further. Nevertheless, even though this may not be the final code to deal with state, better to at least have some decent code in, fixed.